### ISSUE TITLE
Auth code request when refreshing Oauth2 token.

### ISSUE THREAD

**Issue by llrs** [r-lib/httr2#316]
In rtweet, I have some code that uses `oauth_flow_refresh` to manually update the token because the refresh token changes each time it is used (see #186). It was not clear to me at the time the right way to manually manage refresh tokens that change.

But some users are receiving a request to `Enter authorization code:` which seems to come from `oauth_flow_auth_code` (see https://github.com/ropensci/rtweet/issues/779). I don't think this should be happening, but it is not clear to me if I should change something in rtweet or I am misunderstanding how to work with non-typical oauth flows, or this is something you already have in mind. 
Many thanks

**Comment by hadley** [comment]
Can you give me more details? It seems unlikely that you should be using `oauth_flow_refresh()` since that's a flow that really designed for testing, not for end-user use. And the prompt for `Enter authorization code` is only generated by `oauth_flow_auth_code()`.

**Comment by hadley** [comment]
I'm closing this since I haven't heard anything more, but happy to reopen if this is still a problem.

**Comment by llrs** [comment]
I was trying to reproduce it, but got caught with other things. I'll try to provide more info next week. I hope if there is something to be fixed it can be done before the next release which seems near (judging from the activity in the repo).

This is not an issue with the package, at most a request for clearer documentation. As a package developer how should I handle OAuth2 tokens were the refresh token changes each time is used if not with `oauth_flow_refresh()`?

**Comment by hadley** [comment]
As asked, my answer to your question is "you shouldn't"; `oauth_flow_refresh()` is only designed for interactive and testing scenarios. But I suspect your problem is actually that httr2 is missing some functionality that's causing you to try use `oauth_flow_refresh()` this way in the first place.

**Comment by llrs** [comment]
I totally understand that these minors issues are not important right now (or ever). Answer whenever it fits you. 

---

I'm using r-lib/httr2@8f77d41 (For this part of the code it is equal to the , and I have this backtrace of the issue:

```
Backtrace:
    ▆
 1. └─rtweet::rtweet_oauth2()
 2.   └─httr2::oauth_flow_auth_code(...) at rtweet/R/auth.R:493:2
 3.     └─httr2:::oauth_flow_auth_code_read(state)
```

For some months in rtweet I used:

```r
httr2::oauth_flow_auth_code(client,
                       auth_url = "https://twitter.com/i/oauth2/authorize",
                       pkce = TRUE,
                       scope = paste(scopes, collapse = " "),
                       host_name = "127.0.0.1",
                       port = 1410
                       )
```

With the devel version to become 0.3.0 I need to use:

```r
httr2::oauth_flow_auth_code(client,
                       auth_url = "https://twitter.com/i/oauth2/authorize",
                       pkce = TRUE,
                       scope = paste(scopes, collapse = " "),
                       redirect_uri = "http://127.0.0.1:1410")
```

But `normalize_redirect_uri` doesn't detect that as localhost and the `redirect` value returned is `list(uri = "http://127.0.0.1:1410", localhost = FALSE)`. This prevents to run [the branch with](https://github.com/r-lib/httr2/blob/63ef0d6aa31296d18be24fd44757b8eefcc5c01a/R/oauth-flow-auth-code.R#L204) `oauth_flow_auth_code_listen` and triggers the `oauth_flow_auth_code_read` code that asks for 

```
Enter authorization code: 
Enter state parameter:  
```

Which was the poor initial report.

I would suggest to amend `normalize_redirect_uri` to recognize also 127.0.0.1 as localhost [here](https://github.com/r-lib/httr2/blob/63ef0d6aa31296d18be24fd44757b8eefcc5c01a/R/oauth-flow-auth-code.R#L255).

In addition, a related issue prevents the authorization: The `oauth_flow_auth_code_parse` expects the state to match that of `base64_url_rand(32)`. I'm not sure how these challenges work but at least for Twitter it doesn't match. The code returned is a 68 bytes long: I used  `length(base64_url_decode(<code from url>))`, but using some random string of bytes won't match the random string of bytes generated by the authority. 

I am not sure why [this check](https://github.com/r-lib/httr2/blob/63ef0d6aa31296d18be24fd44757b8eefcc5c01a/R/oauth-flow-auth-code.R#L207) is performed. The commit https://github.com/r-lib/httr2/commit/7f603ff5dec12cd6e1e879e479854d002d043b09 where it was introduced doesn't relate it to security but to Rstudio webserver. To me it seems that it shouldn't be there as the proper verification will be done some lines below by `oauth_client_get_token`, at minimum it could be nice to run it only when `is_hosted_session()` is TRUE. 

If manually copied and pasted, `oauth_flow_fetch` later fails with:

```
! OAuth failure [invalid_request]
• Value passed for the redirect uri did not match the uri of the authorization code.
```

The url has encoded the redirect_uri from: `http://127.0.0.1:1410` to: `http%3A%2F%2F127.0.0.1%3A1410` which might cause this.

Tomorrow, I'll open a separate issue about oauth flow for a refresh token changing each time it is used. I hope this time I'll will explain myself better. Many thanks! 

---

BTW, is there a tentative release date ? rtweet is currently failing on CRAN and I would wish to make a new release using httr2 >= 0.2.3.9000 as it could solve several problems and make it viable a release of the package depending on it. (I would ask for an extension if the expected deadline is not too far). Many thanks!

**Comment by hadley** [comment]
* I'll fix the 127.0.0.1 not being recognised as localhost momentarily.
* Can you give me a more complete reprex for the `state` problem in a new issue? I don't think it was introduced in that commit (see e.g. https://github.com/r-lib/httr2/commit/7f603ff5dec12cd6e1e879e479854d002d043b09#diff-49164da1d5d1913e5d079a5abb32c8bdfb514d3907f78fce5f38fe62e95f7696L169). It's a random value that we pass to the API server and it's supposed to pass back unchanged and helps verify that both the request and the response were made by the same person (it's not very important for the typical flow in R).
